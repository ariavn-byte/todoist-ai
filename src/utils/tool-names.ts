/**
 * Centralized tool names module
 *
 * This module provides a single source of truth for all tool names used throughout the codebase.
 * Each tool should import its own name from this module to avoid hardcoded strings.
 * This prevents outdated references when tool names change.
 */
export const ToolNames = {
    // Task management tools
    ADD_TASKS: 'add-tasks',
    COMPLETE_TASKS: 'complete-tasks',
    UPDATE_TASKS: 'update-tasks',
    FIND_TASKS: 'find-tasks',
    FIND_TASKS_BY_DATE: 'find-tasks-by-date',
    FIND_COMPLETED_TASKS: 'find-completed-tasks',

    // Project management tools
    ADD_PROJECTS: 'add-projects',
    UPDATE_PROJECTS: 'update-projects',
    FIND_PROJECTS: 'find-projects',

    // Section management tools
    ADD_SECTIONS: 'add-sections',
    UPDATE_SECTIONS: 'update-sections',
    FIND_SECTIONS: 'find-sections',

    // General tools
    GET_OVERVIEW: 'get-overview',
    DELETE_OBJECT: 'delete-object',
} as const

// Type for all tool names
export type ToolName = (typeof ToolNames)[keyof typeof ToolNames]
