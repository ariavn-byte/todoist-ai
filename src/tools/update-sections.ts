import type { Section } from '@doist/todoist-api-typescript'
import { z } from 'zod'
import { getToolOutput } from '../mcp-helpers.js'
import type { TodoistTool } from '../todoist-tool.js'
import { formatNextSteps } from '../utils/response-builders.js'
import { ToolNames } from '../utils/tool-names.js'

const { FIND_TASKS, GET_OVERVIEW, FIND_SECTIONS } = ToolNames

const SectionUpdateSchema = z.object({
    id: z.string().min(1).describe('The ID of the section to update.'),
    name: z.string().min(1).describe('The new name of the section.'),
})

const ArgsSchema = {
    sections: z.array(SectionUpdateSchema).min(1).describe('The sections to update.'),
}

const updateSections = {
    name: ToolNames.UPDATE_SECTIONS,
    description: 'Update multiple existing sections with new values.',
    parameters: ArgsSchema,
    async execute({ sections }, client) {
        const updatedSections = await Promise.all(
            sections.map((section) => client.updateSection(section.id, { name: section.name })),
        )

        const textContent = generateTextContent({
            sections: updatedSections,
        })

        return getToolOutput({
            textContent,
            structuredContent: {
                sections: updatedSections,
                totalCount: updatedSections.length,
                updatedSectionIds: updatedSections.map((section) => section.id),
            },
        })
    },
} satisfies TodoistTool<typeof ArgsSchema>

function generateNextSteps(sections: Section[]): string[] {
    // Handle empty sections first (early return)
    if (sections.length === 0) {
        return [`Use ${FIND_SECTIONS} to see current sections`]
    }

    // Handle single section case
    if (sections.length === 1) {
        const section = sections[0]
        if (!section) return []

        return [
            `Use ${FIND_TASKS} with sectionId=${section.id} to see existing tasks`,
            `Use ${GET_OVERVIEW} with projectId=${section.projectId} to see project structure`,
            'Consider updating task descriptions if section purpose changed',
        ]
    }

    // Handle multiple sections case
    const projectIds = [...new Set(sections.map((s) => s.projectId))]
    const steps = [`Use ${FIND_SECTIONS} to see all sections with updated names`]

    if (projectIds.length === 1) {
        steps.push(
            `Use ${GET_OVERVIEW} with projectId=${projectIds[0]} to see updated project structure`,
        )
    } else {
        steps.push(`Use ${GET_OVERVIEW} to see updated project structures`)
    }

    steps.push('Consider updating task descriptions if section purposes changed')
    return steps
}

function generateTextContent({
    sections,
}: {
    sections: Section[]
}) {
    const count = sections.length
    const sectionList = sections
        .map((section) => `â€¢ ${section.name} (id=${section.id}, projectId=${section.projectId})`)
        .join('\n')

    const summary = `Updated ${count} section${count === 1 ? '' : 's'}:\n${sectionList}`

    const nextSteps = generateNextSteps(sections)

    const next = formatNextSteps(nextSteps)
    return `${summary}\n${next}`
}

export { updateSections }
