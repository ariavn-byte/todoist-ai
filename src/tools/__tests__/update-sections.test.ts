import type { Section, TodoistApi } from '@doist/todoist-api-typescript'
import { jest } from '@jest/globals'
import {
    createMockSection,
    extractStructuredContent,
    extractTextContent,
} from '../../utils/test-helpers.js'
import { ToolNames } from '../../utils/tool-names.js'
import { updateSections } from '../update-sections.js'

// Mock the Todoist API
const mockTodoistApi = {
    updateSection: jest.fn(),
} as unknown as jest.Mocked<TodoistApi>

const { FIND_TASKS, UPDATE_SECTIONS, GET_OVERVIEW } = ToolNames

describe(`${UPDATE_SECTIONS} tool`, () => {
    beforeEach(() => {
        jest.clearAllMocks()
    })

    describe('updating a single section', () => {
        it('should update a section when id and name are provided', async () => {
            const mockApiResponse: Section = {
                id: 'existing-section-123',
                projectId: '6cfCcrrCFg2xP94Q',
                sectionOrder: 1,
                userId: 'test-user',
                addedAt: '2024-01-01T00:00:00Z',
                updatedAt: '2024-01-01T00:00:00Z',
                archivedAt: null,
                isArchived: false,
                isDeleted: false,
                isCollapsed: false,
                name: 'Updated Section Name',
            }

            mockTodoistApi.updateSection.mockResolvedValue(mockApiResponse)

            const result = await updateSections.execute(
                { sections: [{ id: 'existing-section-123', name: 'Updated Section Name' }] },
                mockTodoistApi,
            )

            expect(mockTodoistApi.updateSection).toHaveBeenCalledWith('existing-section-123', {
                name: 'Updated Section Name',
            })

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Updated 1 section:')
            expect(textContent).toContain(
                'Updated Section Name (id=existing-section-123, projectId=6cfCcrrCFg2xP94Q)',
            )
            expect(textContent).toContain(`Use ${FIND_TASKS} with sectionId=existing-section-123`)

            // Verify structured content
            const structuredContent = extractStructuredContent(result)
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    sections: [mockApiResponse],
                    totalCount: 1,
                    updatedSectionIds: ['existing-section-123'],
                }),
            )
        })
    })

    describe('updating multiple sections', () => {
        it('should update multiple sections and return mapped results', async () => {
            const mockSections = [
                createMockSection({
                    id: 'section-1',
                    projectId: 'project-1',
                    name: 'Updated First Section',
                }),
                createMockSection({
                    id: 'section-2',
                    projectId: 'project-1',
                    name: 'Updated Second Section',
                }),
                createMockSection({
                    id: 'section-3',
                    projectId: 'project-2',
                    name: 'Updated Third Section',
                }),
            ]

            const [section1, section2, section3] = mockSections as [Section, Section, Section]
            mockTodoistApi.updateSection
                .mockResolvedValueOnce(section1)
                .mockResolvedValueOnce(section2)
                .mockResolvedValueOnce(section3)

            const result = await updateSections.execute(
                {
                    sections: [
                        { id: 'section-1', name: 'Updated First Section' },
                        { id: 'section-2', name: 'Updated Second Section' },
                        { id: 'section-3', name: 'Updated Third Section' },
                    ],
                },
                mockTodoistApi,
            )

            // Verify API was called correctly for each section
            expect(mockTodoistApi.updateSection).toHaveBeenCalledTimes(3)
            expect(mockTodoistApi.updateSection).toHaveBeenNthCalledWith(1, 'section-1', {
                name: 'Updated First Section',
            })
            expect(mockTodoistApi.updateSection).toHaveBeenNthCalledWith(2, 'section-2', {
                name: 'Updated Second Section',
            })
            expect(mockTodoistApi.updateSection).toHaveBeenNthCalledWith(3, 'section-3', {
                name: 'Updated Third Section',
            })

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Updated 3 sections:')
            expect(textContent).toContain(
                'Updated First Section (id=section-1, projectId=project-1)',
            )
            expect(textContent).toContain(
                'Updated Second Section (id=section-2, projectId=project-1)',
            )
            expect(textContent).toContain(
                'Updated Third Section (id=section-3, projectId=project-2)',
            )

            // Verify structured content
            const structuredContent = extractStructuredContent(result)
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    sections: mockSections,
                    totalCount: 3,
                    updatedSectionIds: ['section-1', 'section-2', 'section-3'],
                }),
            )
        })

        it('should handle sections from the same project', async () => {
            const mockSections = [
                createMockSection({
                    id: 'section-1',
                    projectId: 'same-project',
                    name: 'Backlog',
                }),
                createMockSection({
                    id: 'section-2',
                    projectId: 'same-project',
                    name: 'Done',
                }),
            ]

            const [section1, section2] = mockSections as [Section, Section]
            mockTodoistApi.updateSection
                .mockResolvedValueOnce(section1)
                .mockResolvedValueOnce(section2)

            const result = await updateSections.execute(
                {
                    sections: [
                        { id: 'section-1', name: 'Backlog' },
                        { id: 'section-2', name: 'Done' },
                    ],
                },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Updated 2 sections:')
            expect(textContent).toContain(`Use ${GET_OVERVIEW} with projectId=same-project`)
        })
    })

    describe('error handling', () => {
        it('should propagate API errors', async () => {
            const apiError = new Error('API Error: Section not found')
            mockTodoistApi.updateSection.mockRejectedValue(apiError)

            await expect(
                updateSections.execute(
                    { sections: [{ id: 'nonexistent', name: 'New Name' }] },
                    mockTodoistApi,
                ),
            ).rejects.toThrow('API Error: Section not found')
        })

        it('should handle partial failures in multiple sections', async () => {
            const mockSection = createMockSection({
                id: 'section-1',
                projectId: 'project-1',
                name: 'Updated Section',
            })

            mockTodoistApi.updateSection
                .mockResolvedValueOnce(mockSection)
                .mockRejectedValueOnce(new Error('API Error: Section not found'))

            await expect(
                updateSections.execute(
                    {
                        sections: [
                            { id: 'section-1', name: 'Updated Section' },
                            { id: 'nonexistent', name: 'New Name' },
                        ],
                    },
                    mockTodoistApi,
                ),
            ).rejects.toThrow('API Error: Section not found')
        })
    })
})
