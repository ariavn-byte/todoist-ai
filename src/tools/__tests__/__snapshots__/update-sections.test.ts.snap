// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`update-sections tool updating a single section should update a section when id and name are provided 1`] = `
"Updated 1 section:
• Updated Section Name (id=existing-section-123, projectId=6cfCcrrCFg2xP94Q)
Next:
- Use find-tasks with sectionId=existing-section-123 to see existing tasks
- Use get-overview with projectId=6cfCcrrCFg2xP94Q to see project structure
- Consider updating task descriptions if section purpose changed"
`;

exports[`update-sections tool updating multiple sections should handle sections from the same project 1`] = `
"Updated 2 sections:
• Backlog (id=section-1, projectId=same-project)
• Done (id=section-2, projectId=same-project)
Next:
- Use find-sections to see all sections with updated names
- Use get-overview with projectId=same-project to see updated project structure
- Consider updating task descriptions if section purposes changed"
`;

exports[`update-sections tool updating multiple sections should update multiple sections and return mapped results 1`] = `
"Updated 3 sections:
• Updated First Section (id=section-1, projectId=project-1)
• Updated Second Section (id=section-2, projectId=project-1)
• Updated Third Section (id=section-3, projectId=project-2)
Next:
- Use find-sections to see all sections with updated names
- Use get-overview to see updated project structures
- Consider updating task descriptions if section purposes changed"
`;
